core/src/main/java/org/elasticsearch/action/fieldstats/FieldStats.java

```diff
@@ -224,352 +240,392 @@ protected void toInnerXContent(XContentBuilder builder) throws IOException {
+                        isSearchable, isAggregatable, formatter, in.readLong(), in.readLong());
+                }
+                return new Date(maxDoc, docCount, sumDocFreq, sumTotalTermFreq,
+                    isSearchable, isAggregatable, formatter);
+
+            case 3:
+                if (hasMinMax) {
+                    return new Text(maxDoc, docCount, sumDocFreq, sumTotalTermFreq,
+                        isSearchable, isAggregatable, in.readBytesRef(), in.readBytesRef());
+                }
+                return new Text(maxDoc, docCount, sumDocFreq, sumTotalTermFreq,
+                        isSearchable, isAggregatable, null, null);
+
+            case 4:
+                InetAddress min = null;
+                InetAddress max = null;
+                if (hasMinMax) {
+                    int l1 = in.readByte();
+                    byte[] b1 = new byte[l1];
+                    int l2 = in.readByte();
+                    byte[] b2 = new byte[l2];
+                    min = InetAddressPoint.decode(b1);
+                    max = InetAddressPoint.decode(b2);
+                }
+                return new Ip(maxDoc, docCount, sumDocFreq, sumTotalTermFreq,
+                    isSearchable, isAggregatable, min, max);
+
+            default:
+                throw new IllegalArgumentException("Unknown type.");
         }
     }
 
-    public static FieldStats read(StreamInput in) throws IOException {
-        FieldStats stats;
-        byte type = in.readByte();
+    public static String typeName(byte type) {
         switch (type) {
             case 0:
-                stats = new Long();
-                break;
+                return "whole-number";
             case 1:
-                stats = new Date();
-                break;
+                return "floating-point";
             case 2:
-                stats = new Double();
-                break;
+                return "date";
             case 3:
-                stats = new Text();
-                break;
+                return "text";
             case 4:
-                stats = new Ip();
-                break;
+                return "ip";
             default:
-                throw new IllegalArgumentException("Illegal type [" + type + "]");
+                throw new IllegalArgumentException("Unknown type.");
         }
-        stats.readFrom(in);
-        return stats;
     }
 
     private final static class Fields {
-
         final static String MAX_DOC = new String("max_doc");
         final static String DOC_COUNT = new String("doc_count");
         final static String DENSITY = new String("density");
         final static String SUM_DOC_FREQ = new String("sum_doc_freq");
         final static String SUM_TOTAL_TERM_FREQ = new String("sum_total_term_freq");
+        final static String IS_SEARCHABLE = new String("is_searchable");
+        final static String IS_AGGREGATABLE = new String("is_aggregatable");
```

Comment:
on the other hand I would agree to keep isSearchable/isAggregatable in the Java API